content/blog/2020-07-01-learning-learnr/index.md

```diff
@@ -0,0 +1,248 @@
+  url: https://unsplash.com/photos/z1c9juteR5c
+---
+
+> "Let us change our traditional attitude to the construction of
+> programs: Instead of imagining that our main task is to instruct a
+> computer what to do, let us concentrate rather on explaining to human
+> beings what we want a computer to do."
+>
+> -- Donald Knuth, *Literate Programming* (1984)
+
+Data scientists engage in and develop tools for literate programming. A
+variant of Markdown called R Markdown has provided the foundation for
+several R packages to assist in writing books ([bookdown](https://bookdown.org/home/)), blogs/websites
+([blogdown](https://bookdown.org/yihui/blogdown/), [hugodown](https://hugodown.r-lib.org)), scientific reports for the web ([distill](https://rstudio.github.io/distill/)) and
+presentations ([xaringan](https://bookdown.org/yihui/rmarkdown/xaringan.html)). The trend with these packages and tools is to
+take the base R Markdown document and *extend* it to provide a powerful
+way to write documents in a simple Markdown format, yet be able to embed
+complex mixed media such as code, images, videos, and interactive
+elements to better suit the type of artifact the user wants to create.
+But, there is yet another level of literate programming that is now
+possible for instructors who want to make interactive lessons thanks to
+the [learnr package](https://rstudio.github.io/learnr/index.html).
+
+For my summer internship, Alison Hill and Greg Wilson initially tasked
+me with building Python lessons using the learnr package with the
+caveat that some infrastructure will need to be built. My journey
+involved a lot of reading and tracing code, as well as consulting Barret
+Schloerke, who provided me both knowledge about the trees and the forest
+of learnr. I started making diagrams to describe the
+architecture and execution flows, and this diagramming activity became a
+wonderful tool to identify my gaps of knowledge. Every time I came back
+to my diagrams, I had new knowledge to incorporate, which again opened
+up new gaps I needed to fill. In this post, I'd like to show
+you what I've learned so far and how I formed a working knowledge of this 
+intricate package; if you'd like to know more about its specific features, please see
+[this post](https://education.rstudio.com/blog/2020/06/summer-camp-hs/) from my fellow intern Ezgi.
+In the end, I hope to convey the big picture of learnr and the
+technologies it combines for interactive lessons.
+
+## What is learnr?
+
+learnr's
+[documentation](https://rstudio.github.io/learnr/index.HTML#Overview)
+says that:
+
+> The learnr package makes it easy to turn any R Markdown document into
+> an interactive tutorial.
+
+For the rest of the post, we'll refer to the following example of a learnr
+document called `addition.Rmd` that only includes an exercise and its hint/checking
+counterparts. Exercise handling is what I am most familiar with so I won't cover
+how questions are handled.
+
+![](./document.jpg)
```

Comment:
I *think* you can always just name the image files like `![](document.jpg)`.

Reply:
Okay, let me do that!